### PR TITLE
update sqlagg

### PR DESCRIPTION
**Pull request by snopoke**
Opening for tests: https://github.com/dimagi/sql-agg/pull/56

### CODE REVIEW

**Review by mkangia — COMMENTED** [review]
Build passed which is great,
This would go after/with the soft asserts i think.

**Comment by snopoke** [comment]
I don't think we would merge this. We'd rather release a new version of sqlagg once we're ready to go and then update the requirements with the new version number

**Comment by mkangia** [comment]
Cool, i was thinking of a different approach. That sounds good @snopoke , thanks.

File: custom/up_nrhm/reports/block_level_month_report.py
Comment (snopoke) [inline custom/up_nrhm/reports/block_level_month_report.py:82]: Based on the columns in this report I think it should only ever return a single row

File: corehq/apps/reports/sqlreport.py
Comment (snopoke) [inline corehq/apps/reports/sqlreport.py:354]: nit: I don't see anywhere that creates a DataFormatter with a row filter so maybe we can just remove that completely? (maybe separate PR)

File: custom/up_nrhm/reports/block_level_month_report.py
Comment (mkangia) [reply custom/up_nrhm/reports/block_level_month_report.py:82]: Still getting a hold of how these reports are working so really don't understand why that would be and comment on that.

Ideally i think all reports should assume receiving back multiple rows with/without group by where the key would be index or group by columns and if a report expects only a single row, it should raise an error if multiple were returned.

File: corehq/apps/reports/sqlreport.py
Comment (mkangia) [reply corehq/apps/reports/sqlreport.py:354]: Did we stop using it? 
https://github.com/dimagi/commcare-hq/pull/871/files#diff-10981343807562b1516015f8cff1d936R233

**Comment by mkangia** [comment]
✔️ 
@snopoke can you please review this and suggest way forward?

File: corehq/apps/reports/sqlreport.py
Comment (snopoke) [reply corehq/apps/reports/sqlreport.py:354]: I'm not sure it was ever used - I don't see any current usages.

File: custom/champ/sqldata.py
Comment (snopoke) [inline custom/champ/sqldata.py:436]: nit

```suggestion
        data = sorted(filter(bool, list(data)))
```

File: custom/up_nrhm/reports/block_level_month_report.py
Comment (snopoke) [reply custom/up_nrhm/reports/block_level_month_report.py:82]: all the columns are aggregate columns (sums or counts) so the SQL should only return a single row

**Review by snopoke — COMMENTED** [review]
Is there a test for a UCR that doesn't have a group by?

File: corehq/apps/reports/sqlreport.py
Comment (mkangia) [reply corehq/apps/reports/sqlreport.py:354]: Cool, will take it up as a followup PR

**Comment by mkangia** [comment]
> Is there a test for a UCR that doesn't have a group by?

@snopoke there are couple of tests in `SimpleReportTest` that dont use group by. Does that cover the case you are thinking?

**Comment by snopoke** [comment]
You can add a test to this class that doesn't have any `aggregation_columns`  e.g.
https://github.com/dimagi/commcare-hq/blob/bd060a7743d26a34a62175e1994fa36c7275efb7/corehq/apps/userreports/tests/test_report_aggregation.py#L41

```python
    def test_aggregation_no_group_by(self):
        report_config = self._create_report(
            aggregation_columns=[],
            columns=[
                {
                    "type": "field",
                    "display": "report_column_display_number",
                    "field": 'indicator_col_id_number',
                    'aggregation': 'max'
                }
            ]
        )
        view = self._create_view(report_config)

        self.assertEqual(
            view.export_table,
            [[
                'foo',
                [
                    ['report_column_display_number'],
                    [4],
                ]
            ]]
        )
```

**Comment by mkangia** [comment]
Hey @czue 
Can you drop a review on this please?

Simon and I have co-contributed to this.

File: corehq/apps/reports/api.py
Comment (czue) [inline corehq/apps/reports/api.py:44]: has this just always been wrong or is it related to the change in some way?

Comment (czue) [reply corehq/apps/reports/api.py:44]: my guess based on the other custom report changes: previously there were some instances that returned a dict if there was only one result that have now been converted to a list (where is this change?)

File: custom/champ/views.py
Comment (czue) [inline custom/champ/views.py:81]: what was the process used to determine that we caught all possible instances of something like this?

File: requirements-python3/dev-requirements.txt
Comment (czue) [inline requirements-python3/dev-requirements.txt:201]: would we not push a proper release before adding it as an HQ dependency? (as opposed to a beta)

File: corehq/apps/reports/api.py
Comment (czue) [reply corehq/apps/reports/api.py:44]: > where is this change?

my guess based on further digging is that's the crux of this change: https://github.com/dimagi/sql-agg/pull/56/

**Review by czue — COMMENTED** [review]
It's a bit hard to wrap my head around this change and the lack of a PR description as well as the one-year-old PR in sqlagg don't really help. But I eventually managed to get big picture sense of what's going on.

No blockers but a few questions. My main concerns are not with the code so much as the code that maybe we didn't write/test.

Also, is this going to impact any existing user-defined UCRs? I have the sense if they didn't have a group_by then the behavior would change, but maybe we don't allow them to do that?

File: custom/champ/views.py
Comment (snopoke) [reply custom/champ/views.py:81]: Other than a thorough code search we've also had a soft assert in place for some time now which has only ever triggered on UCRs (which is addressed in this PR)

File: requirements-python3/dev-requirements.txt
Comment (snopoke) [reply requirements-python3/dev-requirements.txt:201]: indeed - we'll actually need to update this to `0.17.2` since there have been some other changes done since then. @mkangia can you update the version please

File: corehq/apps/reports/api.py
Comment (mkangia) [reply corehq/apps/reports/api.py:44]: this was just to reflect what is expected of classes inheriting from this abstract class, which was indeed not correct. You can see the example in the documentation right above it.

File: custom/champ/views.py
Comment (mkangia) [reply custom/champ/views.py:81]: Plus tests(if they were available for the custom data source/report)

File: requirements-python3/dev-requirements.txt
Comment (mkangia) [reply requirements-python3/dev-requirements.txt:201]: Sure, i will bump that up.

**Review by czue — APPROVED** [review]

**Comment by czue** [comment]
> Also, is this going to impact any existing user-defined UCRs? I have the sense if they didn't have a group_by then the behavior would change, but maybe we don't allow them to do that?

Is there a quick answer to this question? @snopoke @mkangia - I wasn't sure if [this comment](https://github.com/dimagi/commcare-hq/pull/27020#discussion_r443407856) was saying explicitly that it's not a problem

**Comment by mkangia** [comment]
@czue If they don't have a group by, they will start seeing more rows(if that is the correct result) now, instead of just one seen earlier.
fyi @snopoke

**Comment by czue** [comment]
> If they don't have a group by, they will start seeing more rows(if that is the correct result) now, instead of just one seen earlier

thanks @mkangia . don't suppose we know how many reports that might impact or anything, do we? I imagine it's on you all to determine the risk profile of that impact and make a call whether or not to pursue a more cautious rollout.

**Comment by snopoke** [comment]
Re-looking at this. On 'prod' we have:

8533 report configs with no aggregation_columns
443 with data sources that exist
434 with data
12 that don't do any aggregation (all from a single domain)

These 12 reports will now receive more rows. I don't think there is a way to get around that since the current behaviour is incorrect. I will discuss options with the support team.

**Comment by mkangia** [comment]
thank you @snopoke 

> 443 with data sources that exist

this means 443 unique data sources from 8533 reports? or there are reports that don't have an existing data source only?

**Comment by snopoke** [comment]
> this means 443 unique data sources from 8533 reports? or there are reports that don't have an existing data source only?

2nd option - report configs that reference non-existing datasources

**Comment by snopoke** [comment]
Waiting for go-ahead from support team before merging.

**Comment by snopoke** [comment]
Support have notified relevant customers.

**Comment by mkangia** [comment]
:fireworks:

**Pull request merged**
